gnocchi/cli.py

```diff
@@ -228,7 +240,35 @@ def run_watchers():
                       'partitioning. Retrying: %s', e)
             raise tenacity.TryAgain(e)
 
-    def _get_tasks(self):
+        filler = threading.Thread(target=self._fill_sacks_to_process)
+        filler.daemon = True
+        filler.start()
+
+    @utils.retry
+    def _fill_sacks_to_process(self):
+        n = notifier.get_driver(self.conf.notifier)
+        if not n:
+            LOG.warning("Notifier not configured. "
+                        "Metric processing will not be real-time.")
+            # Terminate the thread
+            return
+
+        try:
+            for sack in n.iter_on_sacks_to_process():
+                if sack in self._get_sacks_to_process():
+                    LOG.debug(
+                        "Got notification for sack %s, waking up processing",
+                        sack)
+                    self.sacks_with_measures_to_process.add(sack)
```

Comment:
It is thread-safe because the only shared data structure is a set. Waking-up the main thread is the easiest and "safest" solution.

The discard call is safe, but you're right that another notification can come in the mean time. There's no way to be sure which stage of processing might. There is already a possibility that a sack with N measures is being processed while more measures are being written (remember writing measures is lock-free).

That's why the polling system is kept in place. Notification is just a good way to cut-down the processing interval. There's no good reason to have metricd being sleeping for `metric_processing_delay` when you can know there's jobs to do.